File: processor/src/main/java/io/michaelrocks/lightsaber/processor/graph/DependencyGraph.java
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.graph;

import io.michaelrocks.lightsaber.Injector;
import io.michaelrocks.lightsaber.processor.ProcessingException;
import io.michaelrocks.lightsaber.processor.ProcessorContext;
import io.michaelrocks.lightsaber.processor.descriptors.ModuleDescriptor;
import io.michaelrocks.lightsaber.processor.descriptors.ProviderDescriptor;
import io.michaelrocks.lightsaber.processor.descriptors.QualifiedType;
import org.objectweb.asm.Type;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class DependencyGraph {
    private final Map<QualifiedType, List<QualifiedType>> typeGraph = new HashMap<>();

    public DependencyGraph(final ProcessorContext processorContext, final Collection<ModuleDescriptor> modules) {
        final QualifiedType rootType = new QualifiedType(Type.getType(Injector.class));
        typeGraph.put(rootType, Collections.<QualifiedType>emptyList());
        for (final ModuleDescriptor module : modules) {
            final Set<QualifiedType> providableModuleTypes = new HashSet<>();
            for (final ProviderDescriptor provider : module.getProviders()) {
                final QualifiedType returnType = provider.getQualifiedProvidableType();
                if (providableModuleTypes.add(returnType)) {
                    typeGraph.put(returnType, provider.getDependencies());
                } else {
                    final String message = String.format("Module %s provides %s multiple times",
                            module.getModuleType().getInternalName(), returnType);
                    processorContext.reportError(new ProcessingException(message));
                }
            }
        }
    }

    public Collection<QualifiedType> getTypes() {
        return typeGraph.keySet();
    }

    public Collection<QualifiedType> getTypeDependencies(final QualifiedType type) {
        return typeGraph.get(type);
    }
}

File: processor/src/main/kotlin/io/michaelrocks/lightsaber/processor/graph/DependencyGraph.kt
/*
 * Copyright 2016 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.graph

import io.michaelrocks.grip.mirrors.signature.GenericType
import io.michaelrocks.lightsaber.Injector
import io.michaelrocks.lightsaber.processor.ErrorReporter
import io.michaelrocks.lightsaber.processor.ProcessingException
import io.michaelrocks.lightsaber.processor.model.Dependency
import io.michaelrocks.lightsaber.processor.model.Module
import io.michaelrocks.lightsaber.processor.model.ProvisionPoint
import org.objectweb.asm.Type
import java.util.*

class DependencyGraph(errorReporter: ErrorReporter, modules: Collection<Module>) {
  private val typeGraph = HashMap<Dependency, List<Dependency>>()

  val types: Collection<Dependency>
    get() = typeGraph.keys

  init {
    val rootType = Dependency(GenericType.RawType(Type.getType(Injector::class.java)))
    typeGraph.put(rootType, emptyList<Dependency>())
    for (module in modules) {
      val providableModuleTypes = HashSet<Dependency>()
      for (provider in module.providers) {
        val returnType = provider.dependency
        if (providableModuleTypes.add(returnType)) {
          val injectees =
              (provider.provisionPoint as? ProvisionPoint.AbstractMethod)?.injectionPoint?.injectees.orEmpty()
          val dependencies = injectees.map { it.dependency }
          typeGraph.put(returnType, dependencies)
        } else {
          val message = "Module %s provides %s multiple times".format(module.type.internalName, returnType)
          errorReporter.reportError(ProcessingException(message))
        }
      }
    }
  }

  fun getTypeDependencies(type: Dependency): Collection<Dependency>? = typeGraph[type]
}
